File: himsha-sdk-python/himsha_sdk/transaction.py
"""HIMSHA transaction types: AccountMeta, Instruction, Message, Transaction."""

from __future__ import annotations

import hashlib
import struct
import time
from dataclasses import dataclass, field
from typing import Sequence

from .pubkey import HimshaPublicKey


@dataclass
class AccountMeta:
    pubkey: HimshaPublicKey
    is_signer: bool
    is_writable: bool

    @classmethod
    def writable(cls, pubkey: HimshaPublicKey, is_signer: bool) -> "AccountMeta":
        return cls(pubkey=pubkey, is_signer=is_signer, is_writable=True)

    @classmethod
    def readonly(cls, pubkey: HimshaPublicKey, is_signer: bool) -> "AccountMeta":
        return cls(pubkey=pubkey, is_signer=is_signer, is_writable=False)

    def to_json(self) -> dict:
        return {
            "pubkey":     self.pubkey.to_base58(),
            "isSigner":   self.is_signer,
            "isWritable": self.is_writable,
        }


@dataclass
class HimshaInstruction:
    program_id: HimshaPublicKey
    accounts:   list[AccountMeta]
    data:       bytes

    def to_json(self) -> dict:
        return {
            "programId": self.program_id.to_base58(),
            "accounts":  [a.to_json() for a in self.accounts],
            "data":      self.data.hex(),
        }


@dataclass
class HimshaMessage:
    signers:      list[HimshaPublicKey]
    instructions: list[HimshaInstruction]
    timestamp:    int = field(default_factory=lambda: int(time.time()))

    def hash(self) -> bytes:
        """SHA-256 of the message — this is what signers sign."""
        h = hashlib.sha256()
        for s in self.signers:
            h.update(s.to_bytes())
        for instr in self.instructions:
            h.update(instr.program_id.to_bytes())
            for acc in instr.accounts:
                h.update(acc.pubkey.to_bytes())
                h.update(bytes([1 if acc.is_signer else 0, 1 if acc.is_writable else 0]))
            h.update(instr.data)
        h.update(struct.pack("<Q", self.timestamp))
        return h.digest()

    def to_json(self) -> dict:
        return {
            "signers":      [s.to_base58() for s in self.signers],
            "instructions": [i.to_json() for i in self.instructions],
            "timestamp":    str(self.timestamp),
        }


class HimshaTransaction:
    version = 0

    def __init__(self, message: HimshaMessage) -> None:
        self.message    = message
        self.signatures: list[bytes] = []

    @classmethod
    def create(
        cls,
        signers:      Sequence[HimshaPublicKey],
        instructions: Sequence[HimshaInstruction],
        timestamp:    int | None = None,
    ) -> "HimshaTransaction":
        msg = HimshaMessage(
            signers=list(signers),
            instructions=list(instructions),
            timestamp=timestamp if timestamp is not None else int(time.time()),
        )
        return cls(msg)

    def add_signature(self, sig: bytes) -> "HimshaTransaction":
        if len(sig) != 64:
            raise ValueError(f"Signature must be 64 bytes, got {len(sig)}")
        self.signatures.append(sig)
        return self

    def message_hash(self) -> bytes:
        return self.message.hash()

    def to_json(self) -> dict:
        return {
            "version":    self.version,
            "signatures": [s.hex() for s in self.signatures],
            "message":    self.message.to_json(),
        }

File: himsha-sdk/src/transaction.ts
import { createHash } from 'crypto';
import { HimshaPublicKey } from './pubkey';

export interface AccountMeta {
  pubkey: HimshaPublicKey;
  isSigner: boolean;
  isWritable: boolean;
}

export class HimshaInstruction {
  constructor(
    public readonly programId: HimshaPublicKey,
    public readonly accounts: AccountMeta[],
    public readonly data: Uint8Array,
  ) {}

  static writable(pubkey: HimshaPublicKey, isSigner: boolean): AccountMeta {
    return { pubkey, isSigner, isWritable: true };
  }

  static readonly(pubkey: HimshaPublicKey, isSigner: boolean): AccountMeta {
    return { pubkey, isSigner, isWritable: false };
  }
}

export class HimshaMessage {
  constructor(
    public readonly signers: HimshaPublicKey[],
    public readonly instructions: HimshaInstruction[],
    public readonly timestamp: bigint = BigInt(Math.floor(Date.now() / 1000)),
  ) {}

  /** SHA-256 of the serialized message — this is what each signer signs. */
  hash(): Uint8Array {
    const hasher = createHash('sha256');
    for (const signer of this.signers) {
      hasher.update(signer.toBytes());
    }
    for (const instr of this.instructions) {
      hasher.update(instr.programId.toBytes());
      for (const acc of instr.accounts) {
        hasher.update(acc.pubkey.toBytes());
        hasher.update(new Uint8Array([acc.isSigner ? 1 : 0, acc.isWritable ? 1 : 0]));
      }
      hasher.update(instr.data);
    }
    const tsBytes = new Uint8Array(8);
    new DataView(tsBytes.buffer).setBigUint64(0, this.timestamp, true);
    hasher.update(tsBytes);
    return new Uint8Array(hasher.digest());
  }

  toJSON() {
    return {
      signers:      this.signers.map(s => s.toBase58()),
      instructions: this.instructions.map(instr => ({
        programId: instr.programId.toBase58(),
        accounts:  instr.accounts.map(a => ({
          pubkey:     a.pubkey.toBase58(),
          isSigner:   a.isSigner,
          isWritable: a.isWritable,
        })),
        data: Buffer.from(instr.data).toString('hex'),
      })),
      timestamp: this.timestamp.toString(),
    };
  }
}

export class HimshaTransaction {
  version = 0;
  signatures: Uint8Array[] = [];

  constructor(private message: HimshaMessage) {}

  /** Add an instruction to this transaction (fluent builder). */
  add(instruction: HimshaInstruction): this {
    // create a new message with the added instruction
    (this.message as any).instructions.push(instruction);
    return this;
  }

  /** Sign with a raw 64-byte Schnorr signature. */
  addSignature(sig: Uint8Array): this {
    if (sig.length !== 64) throw new Error('Signature must be 64 bytes');
    this.signatures.push(sig);
    return this;
  }

  messageHash(): Uint8Array {
    return this.message.hash();
  }

  toJSON() {
    return {
      version:    this.version,
      signatures: this.signatures.map(s => Buffer.from(s).toString('hex')),
      message:    this.message.toJSON(),
    };
  }

  static create(
    signers: HimshaPublicKey[],
    instructions: HimshaInstruction[],
    timestamp?: bigint,
  ): HimshaTransaction {
    return new HimshaTransaction(new HimshaMessage(signers, instructions, timestamp));
  }
}
